### PR TITLE
fix(parser): accept unquoted space-concat in field keys per S10.8 (#65)

### LINKED ISSUE #65: Spec violation: S10.8 — unquoted string concat in field key rejected
## Spec reference

HOCON.md L317 / L556: "string value concatenation is allowed in field keys" — unquoted adjacent tokens separated by whitespace must form a single key. Example from L556: \`a b c : 42\` is equivalent to \`\"a b c\" : 42\`.

## Expected behavior

\`\`\`hocon
a b = 1
\`\`\`

should parse as key \`\"a b\"\` with value \`1\`.

## Actual behavior

\`\`\`go
ParseString(\"a b = 1\")
// → parse error at line 1, col 3: expected ':', '=' or '{' after key
\`\`\`

The lexer/parser treats \`a\` as the key and then expects an immediate separator, rejecting the second token.

## Fix area

\`internal/parser/parser.go\` — \`parseFieldKey\` or equivalent: should accept adjacent unquoted tokens separated by whitespace as a single concatenated key.

## Test

Pinned by \`t.Skipf\` in \`internal/parser/parser_test.go\` (\`TestSpecS10_8_UnquotedConcatInKey\`).

### PR DESCRIPTION
**Pull request by y1o1**
## Summary

- Implements S10.8 (HOCON spec L317 + L553-560) — `foo bar = 1` now parses as key `["foo bar"]` instead of erroring with `expected ':', '=' or '{' after key`.
- `parseKey` gains a `spaceConcat` flag: when the next key token has `PrecedingSpace=true`, the first dot-split piece merges into the LAST existing segment with a literal space; any remaining pieces become new path segments.
- Leading `.` after whitespace keeps path-separator semantics per S11.1 — handled incidentally via the pre-existing leading-dot continuation branch (which fires before space-concat); a defensive guard in the space-concat merge path covers any future reordering.
- 12 regression tests pin: basic 2-token, 3-token spec L556 example, dotted-prefix + dotted-tail concat, quoted+unquoted concat, inline-object shorthand, 4 leading-dot edge cases, and tab whitespace. The existing `t.Skipf` pin is converted to positive assertion.

## Behaviour examples (all pinned by tests)

| Input | Key (post-fix) |
|---|---|
| `foo bar = 1` | `["foo bar"]` |
| `a b c : 42` | `["a b c"]` (spec L556) |
| `a.b c = 1` | `["a", "b c"]` |
| `a b.c = 1` | `["a b", "c"]` |
| `"foo bar" baz = 1` | `["foo bar baz"]` |
| `a b { x = 1 }` | key `["a b"]`, value the inline object |
| `a .b = 1` | `["a", "b"]` (leading dot stays separator per S11.1) |
| `a .b.c = 1` | `["a", "b", "c"]` |
| `"a" .b = 1` | `["a", "b"]` |
| `a.b .c = 1` | `["a", "b", "c"]` |
| `"a.b" .c = 1` | `["a.b", "c"]` |
| `a\tb = 1` | `["a b"]` (tab counts as whitespace) |

## Cross-impl

- [x] ts.hocon [#128](https://github.com/o3co/ts.hocon/pull/128) — MERGED (1fc0582)
- [x] rs.hocon [#115](https://github.com/o3co/rs.hocon/pull/115) — MERGED (ebb06f4)
- [ ] **this PR (go.hocon #65)**

Closes the v1.5.0 S10.8 scope. xx.hocon compliance-matrix update will be a separate PR after this lands.

## Known limitation (out of scope)

S8.6 (leading-`-` rule) is enforced strictly in key position across all 3 impls (rs/ts/go), but Lightbend's typesafe-config 1.4.3 is **more permissive** — accepts `foo -bar = 1` and similar. This was discovered via FCoT during rs.hocon #115 review (direct typesafe-config JVM probe). Out of scope for the S10.8 v1.5.0 work; tracked as [xx.hocon#42](https://github.com/o3co/xx.hocon/issues/42) for a cross-impl follow-up wave after v1.5.0 ships.

Closes #65.

## Test plan

- [x] `go test ./...` — all packages pass (config, lexer, parser, properties, resolver)
- [x] `go vet ./...` — clean
- [x] `golangci-lint run ./internal/parser/...` — 0 issues
- [x] All 13 S10.8 tests pass (1 baseline + 12 new)
- [x] No regression on existing key tests (numeric+concat for `123abc` key style still works; `a.b` dotted-path still works; `a."b.c".d` mixed quoted+unquoted still works)
- [x] Value-position concat (`x = foo bar`) unaffected — change is localised to `parseKey`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/parser/parser.go [hunk 1]
```diff
@@ -605,6 +605,21 @@ func (p *parser) parseKey() ([]string, error) {
...
+	//   `a b = 1`     → ['a b']
+	//   `a b c : 42`  → ['a b c']            (spec L556 example)
+	//   `a.b c = 1`   → ['a', 'b c']         (concat into last segment)
+	//   `"a" b = 1`   → ['a b']              (quoted + unquoted)
+	// Leading '.' after whitespace already keeps separator semantics via the
+	// existing leading-dot continuation check (which fires before this new
+	// space-concat branch and ignores PrecedingSpace), so `a .b = 1` → ['a', 'b']
+	// works incidentally — the leading-dot path takes priority over space-concat.
+	// Newlines break the chain (S10.7): the lexer emits a separator token
+	// which falls through to the loop's else / break.
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser.go]: Fixed in 1121dd4 — rephrased to name `TokenNewline` explicitly and explain the falls-through-switch mechanism. The token is not a separator; it's the newline token kind that doesn't appear in the space-concat switch case list, so the continuation check fails and the loop breaks.

File: internal/parser/parser_test.go [hunk 2]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
 // adjacent tokens separated by whitespace form a single concatenated key.
 // Per HOCON L317/L556: `a b c : 42` is equivalent to `"a b c" : 42`.
-// Status: ❌ — parser rejects unquoted multi-token keys (see #65).
+// Status: ✅ — fixed in #65; parseKey accepts space-concat continuations.
 func TestSpecS10_8_UnquotedConcatInKey(t *testing.T) {
-	t.Skipf("spec violation, see #65")
 	// Spec example (L556): `a b c : 42` must parse as key "a b c" → 42.
 	obj, err := parser.Parse(`a b = 1`)
 	if err != nil {
 		t.Fatalf("spec L317/L556: 'a b = 1' must be accepted as key 'a b', got error: %v", err)
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — removed the stale L556 spec example mention from the basic 2-token test. The L556 verbatim example (`a b c : 42`) lives in its own `TestSpecS10_8_ThreeTokenSpecExample`. The basic test now cites the more general L317 (string

File: internal/parser/parser_test.go [hunk 3]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
-	if len(obj.Fields) != 1 || obj.Fields[0].Key[0] != "a b" {
-		t.Errorf("expected key ['a b'], got %v", func() interface{} {
-			if len(obj.Fields) > 0 {
-				return obj.Fields[0].Key
-			}
-			return nil
-		}())
+	if len(obj.Fields) != 1 || len(obj.Fields[0].Key) != 1 || obj.Fields[0].Key[0] != "a b" {
+		t.Errorf("expected key ['a b'], got %v", obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 4]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+
+func TestSpecS10_8_SpaceConcatAfterDottedPath(t *testing.T) {
+	// `a.b c = 1` → ['a', 'b c']: concat merges into the LAST path segment.
+	obj, err := parser.Parse(`a.b c = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a", "b c"}
+	if len(obj.Fields[0].Key) != 2 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 5]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+func TestSpecS10_8_SpaceConcatDottedTail(t *testing.T) {
+	// `a b.c = 1` → ['a b', 'c']: head merges, tail pushes as new segment.
+	obj, err := parser.Parse(`a b.c = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a b", "c"}
+	if len(obj.Fields[0].Key) != 2 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 6]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+
+func TestSpecS10_8_QuotedThenUnquotedSpaceConcat(t *testing.T) {
+	// `"foo bar" baz = 1` → ['foo bar baz']: quoted segment + unquoted tail via S10.8.
+	obj, err := parser.Parse(`"foo bar" baz = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if len(obj.Fields[0].Key) != 1 || obj.Fields[0].Key[0] != "foo bar baz" {
+		t.Errorf("expected key ['foo bar baz'], got %v", obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 7]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+
+func TestSpecS10_8_InlineObjectShorthand(t *testing.T) {
+	// `a b { x = 1 }` — key ['a b'], value the inline object.
+	obj, err := parser.Parse(`a b { x = 1 }`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if len(obj.Fields[0].Key) != 1 || obj.Fields[0].Key[0] != "a b" {
+		t.Errorf("expected key ['a b'], got %v", obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 8]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+	// (3-way convergence finding from ts.hocon #128: Claude+Codex+Copilot all
+	// flagged this on the ts companion PR; rs.hocon #115 ported the same fix.)
+	obj, err := parser.Parse(`a .b = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a", "b"}
+	if len(obj.Fields[0].Key) != 2 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 9]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+func TestSpecS10_8_LeadingDotAfterWhitespaceMultiSegmentTail(t *testing.T) {
+	// `a .b.c = 1` → ['a', 'b', 'c'].
+	obj, err := parser.Parse(`a .b.c = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a", "b", "c"}
+	if len(obj.Fields[0].Key) != 3 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] || obj.Fields[0].Key[2] != want[2] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 10]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+func TestSpecS10_8_LeadingDotAfterQuotedThenWhitespace(t *testing.T) {
+	// `"a" .b = 1` → ['a', 'b'].
+	obj, err := parser.Parse(`"a" .b = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a", "b"}
+	if len(obj.Fields[0].Key) != 2 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 11]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+func TestSpecS10_8_DottedPathThenWhitespaceThenDot(t *testing.T) {
+	// `a.b .c = 1` → ['a', 'b', 'c'] (Copilot-flagged convergence case on ts).
+	obj, err := parser.Parse(`a.b .c = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a", "b", "c"}
+	if len(obj.Fields[0].Key) != 3 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] || obj.Fields[0].Key[2] != want[2] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 12]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+	// `"a.b" .c = 1` → ['a.b', 'c']: quoted-key literal 'a.b' preserved (no
+	// path-split inside quoted), and `.c` becomes a sibling segment.
+	obj, err := parser.Parse(`"a.b" .c = 1`)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	want := []string{"a.b", "c"}
+	if len(obj.Fields[0].Key) != 2 || obj.Fields[0].Key[0] != want[0] || obj.Fields[0].Key[1] != want[1] {
+		t.Errorf("expected key %v, got %v", want, obj.Fields[0].Key)
+	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.

File: internal/parser/parser_test.go [hunk 13]
```diff
@@ -592,21 +592,148 @@ func TestSpecS10_8_QuotedKeyWithSpaceAllowed(t *testing.T) {
...
+func TestSpecS10_8_TabBetweenKeyTokensIsSpaceConcat(t *testing.T) {
+	// `PrecedingSpace` covers any HOCON whitespace; the merge joiner is
+	// canonical U+0020 (per S10.6 whitespace normalisation).
+	obj, err := parser.Parse("a\tb = 1")
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if len(obj.Fields[0].Key) != 1 || obj.Fields[0].Key[0] != "a b" {
+		t.Errorf("expected key ['a b'], got %v", obj.Fields[0].Key)
 	}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/parser/parser_test.go]: Fixed in 1121dd4 — centralised via a new `assertKeyPath` helper that first asserts `len(obj.Fields) == 1` (via `t.Fatalf`) before per-segment comparison. Same finding flagged on 11 tests → one helper closes all of them and removes the panic risk on every site. Per-test indexing into `obj.Fields[0]` is eliminated.
